src/Tools/Auditor/Report.dfy

```diff
@@ -0,0 +1,125 @@
+include "../../Utils/Library.dfy"
+
+module AuditReport {
+  import opened Utils.Lib.Seq
+
+  //// Data types for report ////
+
+  datatype Tag =
+    | IsGhost
+    | IsSubsetType
+    | IsCallable
+    | HasNoBody // TODO: should all tags be positive?
+    | HasAxiomAttribute
+    | HasExternAttribute
+    | HasVerifyFalseAttribute
+    | HasAssumeInBody // Maybe: (expr: Expression) TODO: :axiom?
+    | HasRequiresClause
+    | HasEnsuresClause
+    | HasNoWitness // TODO: should all tags be positive?
+    | HasJustification
+    // TODO: decreases *?
+    {
+      function method ToString(): string {
+        match this {
+          case IsGhost => "IsGhost"
+          case IsSubsetType => "IsSubsetType"
+          case IsCallable => "IsCallable"
+          case HasNoBody => "HasNoBody"
+          case HasAxiomAttribute => "HasAxiomAttribute"
+          case HasExternAttribute => "HasExternAttribute"
+          case HasVerifyFalseAttribute => "HasVerifyFalseAttribute"
+          case HasAssumeInBody => "HasAssumeInBody"
+          case HasRequiresClause => "HasRequiresClause"
+          case HasEnsuresClause => "HasEnsuresClause"
+          case HasNoWitness => "HasNoWitness"
+          case HasJustification => "HasJustification"
+        }
+      }
+    }
+
+  datatype Assumption =
+    Assumption(name: string, tags: set<Tag>)
+
+  datatype Report =
+    Report(assumptions: seq<Assumption>)
+
+  const EmptyReport := Report([])
+
+  //// Tag categorization ////
+
+  predicate method IsAssumption(ts: set<Tag>) {
+    || (IsSubsetType in ts && HasNoWitness in ts)
+    || HasAxiomAttribute in ts
+    || (&& IsCallable in ts
+        && (|| (HasEnsuresClause in ts && (HasNoBody in ts || HasExternAttribute in ts))
```

Comment:
An extern doesn't need an ensures clause to be unsound; e.g. an extern could return a value of an empty type.

Reply:
Yeah, very true. Do you know how easy it is within the Dafny implementation as-is to identify whether a type might be empty?

Reply:
There's a property on types that indicates whether they are inhabited; I'm not sure if it's computed correctly for subset types.  Let's create an issue about this.

Reply:
#18